queries/bus.py

```diff
@@ -301,81 +257,151 @@ def QBusWord(node: Node, params: QueryStateDict, result: Result) -> None:
+    for speech synthesis.
+    """
+    if stop.endswith(" N") and straeto.BusStop.named(stop[:-1] + "S"):
+        stop = stop[:-1] + "norður"
+    elif stop.endswith(" A") and straeto.BusStop.named(stop[:-1] + "V"):
+        stop = stop[:-1] + "austur"
+    elif stop.endswith(" S") and straeto.BusStop.named(stop[:-1] + "N"):
+        stop = stop[:-1] + "suður"
+    elif stop.endswith(" V") and straeto.BusStop.named(stop[:-1] + "A"):
+        stop = stop[:-1] + "vestur"
+    return stop
+
+
+_BETTER_NAMES = {
+    "BSÍ": "Umferðarmiðstöðin",
+    "FMOS": "Framhaldsskólinn í Mosfellsbæ",
+    "KEF - Airport": "Keflavíkurflugvöllur",
+    "KR": "Knattspyrnufélag Reykjavíkur",
+    "RÚV": "Útvarpshúsið",
+}
+_ABBREV_RE = re.compile(r"\b[A-ZÁÐÉÍÓÚÝÞÆÖ]+\b")
+
+
+def voicify_stop_name(np: str) -> str:
+    """Fix stop name to better suit speech synthesis."""
+    np = _compass_directions(np)
+    np = _BETTER_NAMES.get(np, np)
+    return _ABBREV_RE.sub(lambda m: gssml(m.group(0), type="spell"), np)
 
 
 @lru_cache(maxsize=None)
-def to_accusative(np: str) -> str:
-    """Return the noun phrase after casting it from nominative to accusative case"""
-    np = straeto.BusStop.voice(np)
-    return queries.to_accusative(np, filter_func=_filter_func)
+def accusative_form(np: str, voice: bool = False) -> str:
+    """
+    Return accusative case of the stop name,
+    optionally expanding abbreviations for speech synthesis.
+    """
+    if voice:
+        np = _compass_directions(np)
+        np = _BETTER_NAMES.get(np, np)
+
+    split_symb = None
+    if " / " in np:
+        split_symb = " / "
+    elif " - " in np:
+        split_symb = " - "
+
+    if split_symb is None:
+        # Bus stop is single noun phrase
+        np = NounPhrase(np).accusative or np
+    else:
+        # Bus stop consists of two (at least) noun phrases
+        # separated by split_symb, inflect them separately
+        new_np: List[str] = []
+        for n in np.split(split_symb):
+            if any(c.islower() for c in n):
+                # Not an all-uppercase abbreviation, try to inflect it
+                n = NounPhrase(n).accusative or n
+            new_np.append(n)
+        np = split_symb.join(new_np)
+    if voice:
+        np = _ABBREV_RE.sub(lambda m: gssml(m.group(0), type="spell"), np)
+    return np
 
 
 @lru_cache(maxsize=None)
-def to_dative(np: str) -> str:
-    """Return the noun phrase after casting it from nominative to dative case"""
-    np = straeto.BusStop.voice(np)
-    return queries.to_dative(np, filter_func=_filter_func)
+def dative_form(np: str, voice: bool = False) -> str:
+    """
+    Return dative case of the stop name,
+    optionally expanding abbreviations for speech synthesis.
+    """
+    if voice:
+        np = _compass_directions(np)
+        np = _BETTER_NAMES.get(np, np)
+
+    split_symb = None
+    if " / " in np:
+        split_symb = " / "
+    elif " - " in np:
+        split_symb = " - "
+
+    if split_symb is None:
```

Comment:
Isn't this logic also duplicated? Maybe add an abstract fn to handle this split handling logic